scene/2d/tile_map.cpp

```diff
@@ -3070,18 +3070,20 @@ void TileMap::_build_runtime_update_tile_data(SelfList<TileMapQuadrant>::List &r
 						continue;
 					}
 
-					TileSetAtlasSource *atlas_source = Object::cast_to<TileSetAtlasSource>(source);
-					if (atlas_source) {
-						bool ret = false;
-						if (GDVIRTUAL_CALL(_use_tile_data_runtime_update, q.layer, E_cell.value, ret) && ret) {
+					bool ret = false;
+					if (GDVIRTUAL_CALL(_use_tile_data_runtime_update, q.layer, E_cell.value, ret) && ret) {
+						TileSetAtlasSource *atlas_source = Object::cast_to<TileSetAtlasSource>(source);
+						if (atlas_source) {
 							TileData *tile_data = atlas_source->get_tile_data(c.get_atlas_coords(), c.alternative_tile);
 
-							// Create the runtime TileData.
 							TileData *tile_data_runtime_use = tile_data->duplicate();
 							tile_data->set_allow_transform(true);
 							q.runtime_tile_data_cache[E_cell.value] = tile_data_runtime_use;
 
 							GDVIRTUAL_CALL(_tile_data_runtime_update, q.layer, E_cell.value, tile_data_runtime_use);
+						} else {
+							// We don't have runtime data for this tile. Make sure it's cleared out.
+							q.runtime_tile_data_cache.erase(E_cell.value);
```

Comment:
I don't think that's a proper fix. Depending on the user's custom logic it's still possible for the `GDVIRTUAL_CALL(_use_tile_data_runtime_update, q.layer, E_cell.value, ret) && re` to return `true` and later `false` for the same cell coords containing a tile from TileSetAtlasSource. Meaning the initially added cache entry for this cell would still be in the cache even though the value (TileData) would be deleted. Hence it could likely still crash (haven't actually tested this).

A proper fix I think would be to delete this part as it's not needed here:
```suggestion
```
and make sure the cache is cleared after the deletions. Here:
https://github.com/godotengine/godot/blob/24cb43a8741c7b10abbbbc77bb6e2bc188662ce0/scene/2d/tile_map.cpp#L922-L930
after the `for` loop there should be `dirty_quadrant_list.first()->self()->runtime_tile_data_cache.clear()` added (+ potentially simplify this by factoring out the quadrant or cache into a variable).